File: Week01/main.py
import zlib


def crc32_table(poly: int) -> list:
    crc_table = []
    for i in range(256):
        crc = i;
        for _ in range(8):
            if (crc & 1):
                crc = (crc >> 1) ^ poly
            else:
                crc = crc >> 1
        crc_table.append(crc)

    return crc_table

def crc32_genPoly() -> int:
    # G(x) = x^32 + x^26 + x^23 + x^22 + x^16 + x^12 + x^11 +x^10 + x^8 + x^7 + x^5 + x^4 + x^2 + x + 1
    exponents = [32, 26, 23, 22, 16, 12, 11, 10, 8, 7, 5, 4, 2, 1, 0]

    # from wiki
    # G(x) = x^31 + x^30 + x^26 + x^25 + x^24 + x^18 + x^15 + x^14 + x^12 + x^11 + x^10 + x^8 + x^6 + x^5 + x^4 + x^3 + x + 1
    # exponents = [31, 30, 26, 25, 24, 18, 15, 14, 12, 11, 10, 8, 6, 5, 4, 3, 1, 0]
    
    # 计算多项式值
    polynomial = sum(1 << exp for exp in exponents)
    # result is |32|31|30|29|28|27|26|25|24|23|22|21|20|19|18|17|16|15|14|13|12|11|10|09|08|07|06|05|04|03|02|01|00|
    #           | 1| 0| 0| 0| 0| 0| 1| 0| 0| 1| 1| 0| 0| 0| 0| 0| 1| 0| 0| 0| 1| 1| 1| 0| 1| 1| 0| 1| 1| 0| 1| 1| 1|
    # 0x104C11DB7
    
    return polynomial & 0xFFFFFFFF

def crc32_genPoly_reverse(poly: int) -> int:
    
    reversed_poly = 0

    for i in range(32):
        if poly & (1 << i):
            reversed_poly |= 1 << (31 - i)

    return reversed_poly

def crc32(data: bytes, genx: int) -> int:
    # Initial start with 0xFFFFFFFF
    crc = 0xFFFFFFFF

    for byte in data:
        # crc XOR databyte
        crc ^= byte
        for _ in range(8):
            # Mod 2
            # if crc result last bit is 1 , crc shift right and XOR with gen(x)
            # if is 0, crc shift right only
            if (crc & 1):
                crc = (crc >> 1) ^ genx
            else:
                crc >>= 1
        print(f"byte:{byte}, crc:{crc:08X}")
    
    return crc ^ 0xFFFFFFFF

def crc32_zlib(data: bytes) -> int:
    return zlib.crc32(data)

def crc32_by_table(data: bytes, table: list) -> int:
    crc = 0xFFFFFFFF

    for byte in data:
        idx = (crc ^ byte) & 0xFF
        crc = table[idx] ^ (crc >> 8)
    return crc ^ 0xFFFFFFFF

def main():
    genx = crc32_genPoly()
    genx_r = crc32_genPoly_reverse(genx)
    print(f"0x{genx:08X}, {genx:032b}")
    print(f"0x{genx_r:08X}, {genx_r:32b}")

    test_data = b"ABC"

    crc = crc32(test_data, genx_r)
    print(f"crc:0x{crc: 08X}")

    crczlib = crc32_zlib(test_data)
    print(f"crc32_zlib:0x{crczlib: 08X}")


    table = crc32_table(genx_r)
    # print(table)
    crc_t = crc32_by_table(test_data, table);
    print(f"crct:0x{crc_t: 08X}")

    return



if __name__ == '__main__':
    main()

File: Week01/main.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>

void print_binary(uint32_t data){
	printf("Binary: ");
    for (int i = 31; i >= 0; i--) {
        printf("%d", (data >> i) & 1);
    }
    printf("\n");	
}

int init_crc32_table(uint32_t* table, uint32_t poly){
	int i, j;
	uint32_t crc;

	for (i = 0; i < 256; i++){
		crc = i;
		for (j = 0; j < 8; j++){
			if (crc & 1){
				crc = crc >> 1 ^ poly;
			}else{
				crc = crc >> 1;
			}
		}
		table[i] = crc;
	}

	return 0;
}


uint32_t genX_formula(void){
	uint32_t genX = 0;
	int exp_list[] = {32, 26, 23, 22, 16, 12, 11, 10, 8, 7, 5, 4, 2, 1, 0};

	int i;
	int list_len = sizeof(exp_list)/sizeof(exp_list[0]);
	printf("list_len: %d\n", list_len);

	for (i = 0; i < list_len; i++){
		genX = genX | (1 << exp_list[i]);
	}

	return genX;
}

uint32_t genX_formula_revert(uint32_t genX){

	uint32_t genX_r = 0;
	int i;

	for (i = 0 ; i < 32; i++){
		if (genX & (1 << i)){
			genX_r = genX_r | (1 << (31 - i));
		}
	}

	return genX_r;
}

uint32_t crc32(const uint8_t* data, int len, uint32_t poly){
	uint32_t crc = 0xFFFFFFFF;
	int i, j;

	for(i = 0; i < len; i++){
		crc ^= data[i];
		for (j = 0 ; j < 8; j++){
			if (crc & 1){
				crc = (crc >> 1) ^ poly;
			}else{
				crc = crc >> 1;
			}
		}
		printf("data:[%d], crc:[0x%8X]\n", data[i], crc);
	}

	return crc ^ 0xFFFFFFFF; // revert
}

uint32_t crc32_by_table(const uint8_t* data, int len, uint32_t *table){
	uint32_t crc = 0xFFFFFFFF;
	int i;
	uint8_t idx;

	for (i = 0 ; i < len; i++){
		idx = (crc ^ data[i]) & 0xFF;
		crc = table[idx] ^ (crc >> 8);
	}

	return crc ^ 0xFFFFFFFF;
}


int main(int argc, char const *argv[])
{
	/* code */

	uint32_t genX; 
	uint32_t genX_r;
	uint8_t test_data[] = "ABC";
	int data_len = sizeof(test_data)/sizeof(test_data[0]) - 1;
	uint32_t crc_result;
	uint32_t crc_table[256];
	uint32_t crc_result_table;

	genX = genX_formula();
	genX_r = genX_formula_revert(genX);
	printf("genX: 0x%8X\n", genX);
	print_binary(genX);
	printf("genX_r: 0x%8X\n", genX_r);
	print_binary(genX_r);
	printf("data_len:%d\n", data_len);
	crc_result = crc32(test_data, data_len, genX_r);
	printf("crc: 0x%8X\n", crc_result);

	init_crc32_table(crc_table, genX_r);
	printf("crc_table[1]:0x%8X, crc_table[2]:0x%8X\n", crc_table[1], crc_table[2]);

	crc_result_table = crc32_by_table(test_data, data_len, crc_table);
	printf("crc32_by_table: 0x%8X\n", crc_result_table);

	return 0;
}

